src/Storages/TimeSeries/TimeSeriesInnerTablesCreator.cpp

```diff
@@ -93,6 +96,26 @@ namespace
                     columns.add(std::move(all_tags_column));
                 }
 
+                /// Columns "min_time" and "max_time".
+                if (time_series_settings.store_min_time_and_max_time)
```

Comment:
AggregatingMergeTree tables currently, not as good as RMT with FINAL performance, 
It's not a problem, if we want to keep only one of those metrics, max or min, because we can easily keep only one of version: first or last in RMT, but if we want to keep both at the same time...

Just some idea on how it can be optimized

What if we can have 2 rows representations for each tag combination in tables:

version UInt32 DEFAULT if(direction = 1, ts, MAX_UINT32 - ts)

ENGINE=ReplacingMergeTree(version)
PARTITION BY (direction)

But, it also does mean that we will have double write in that table and make it twice as big, (If we wouldn't do any kind of optimizations for that, like remembering which tag combination we already saw, to not write min_timestamp row)

It does mean that, it may allow us to optimize queries (get list of timeseries, which exist in specific timeframe):

```
SELECT * FROM tags_replacing FINAL WHERE (timestamp < max_timestamp) AND (direction = 1) 
INTERSECT
SELECT * FROM tags_replacing FINAL WHERE (timestamp > min_timestamp) AND (direction = -1) 
```
 
Or make ClickHouse smart enough, to handle :

```
SELECT * FROM tags_aggregating FINAL PREWHERE max_timestamp > min_timestamp_literal AND  min_timestamp < max_timestamp_literal WHERE (max_timestamp < max_timestamp_literal) AND  (min_timestamp > min_timestamp_literal) 
```

So, when ClickHouse execute PREWHERE before FINAL, it may filter out tags/rows (and their ORDER BY key combination), which by definition doesn't match our conditions, and not aggregate them at all at FINAL step.

Reply:
`SELECT FINAL` is not used, that aggregation is used during merges only. The query which is used to read times series is [here](https://github.com/ClickHouse/ClickHouse/blob/85d1aaf5a7e750da0fcd8b6becfa142294b3f47e/src/Storages/TimeSeries/PrometheusRemoteReadProtocol.cpp#L168).

Reply:
I'm not meant "expression" queries, 

I assumed, it's for https://prometheus.io/docs/prometheus/latest/querying/api/#finding-series-by-label-matchers